### PR TITLE
tools: Add scsilatency, scsisnoop, f2fsdist, f2fsslower tools.

### PR DESCRIPTION
**Pull request by weibang6liu**
We used BCC to monitor I/O stack performance on mobile(AArch64) platform, and found it's very efficient and use easily. Meanwhile, we found lack of some tools in multiple layers，then we devloped 4 tools for SCSI layer and F2FS layer.
We hope these tools can help more people.

### CODE REVIEW

File: tools/f2fsdist.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    char *op = "open";
+    return trace_return(ctx,op);
+}
+int trace_fsync_return(struct pt_regs *ctx)
+{
+    char *op = "fsync";
+    return trace_return(ctx,op);
+}
+"""
+# Starting from Linux 4.10 f2fs_file_operations.read_iter has been changed from
```
Comment (netedwardwu) [inline tools/f2fsdist.py:137]: There are many f2fs tracepoints, not sure if those can help.

File: tools/f2fsdist.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    // f2fs filter on file->f_op == f2fs_file_operations
+    struct file *fp = iocb->ki_filp;
+    if ((u64)fp->f_op != %s)
+        return 0;
+
+    u64 ts = bpf_ktime_get_ns();
+    start.update(&tid, &ts);
+    return 0;
+}""" % f2fs_file_ops_addr
```
Comment (netedwardwu) [inline tools/f2fsdist.py:183]: f2fs_file_ops_addr is an address in kallsyms, what if the platform's kptr_restrict is 2?

File: tools/f2fsslower.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+"""
+
+# code replacements
+with open(kallsyms) as syms:
+    ops = ''
+    for line in syms:
+        (addr, size, name) = line.rstrip().split(" ", 2)
+        name = name.split("\t")[0]
+        if name == "f2fs_file_operations":
+            ops = "0x" + addr
```
Comment (netedwardwu) [inline tools/f2fsslower.py:233]: f2fs_file_ops_addr is an address in kallsyms, what if the platform's kptr_restrict is 2?

File: tools/f2fsslower_example.txt [hunk 4]
```diff
@@ -0,0 +1,260 @@
+Demonstrations of f2fsslower, the Linux eBPF/bcc version.
+
+
+f2fsslower shows f2fs reads, writes, opens, and fsyncs, slower than a threshold.
+For example:
+
+# ./f2fsslower
+Tracing f2fs operations slower than 10 ms
+TIME     COMM           PID    T BYTES   OFF_KB   LAT(ms) FILENAME
+23:27:14 Thread-11      13086  R 33554432 0         123.24 0
```
Comment (netedwardwu) [inline tools/f2fsslower_example.txt:10]: Alignment for LAT(ms) & FILENAME?

File: tools/f2fsslower_example.txt [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+optional arguments:
+  -h, --help         show this help message and exit
+  -j, --csv          just print fields: comma-separated values
+  -p PID, --pid PID  trace this PID only
+
+examples:
+    ./f2fsslower             # trace operations slower than 10 ms (default)
+    ./f2fsslower 1           # trace operations slower than 1 ms
+    ./f2fsslower -j 1        # ... 1 ms, parsable output (csv)
```
Comment (netedwardwu) [inline tools/f2fsslower_example.txt:258]: ... ?

File: tools/scsilatency_example.txt [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+       256 -> 511        : 436      |                                        |
+       512 -> 1023       : 205      |                                        |
+      1024 -> 2047       : 383      |                                        |
+      2048 -> 4095       : 759      |                                        |
+      4096 -> 8191       : 2775     |*                                       |
+      8192 -> 16383      : 1067     |                                        |
+     16384 -> 32767      : 2309     |*                                       |
+     32768 -> 65535      : 97       |                                        |
+
+This output sows that only sda device has be accessed, usually between 0.03 ms
```
Comment (netedwardwu) [inline tools/scsilatency_example.txt:105]: typo

File: tools/scsisnoop.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        infobyreq.update(&req, &val);
+    }
+    return 0;
+}
+
+// time SCSI I/O
+int trace_req_start(struct pt_regs *ctx, struct scsi_cmnd *cmd)
+{
+    struct request *req = cmd->request;
+    //bpf_trace_printk("[Andy]trace_start flag: %x\\n", req->cmd_flags);
```
Comment (netedwardwu) [inline tools/scsisnoop.py:86]: ?

File: tools/scsisnoop.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+// output
+int trace_req_completion(struct pt_regs *ctx, struct scsi_cmnd *cmd)
+{
+    struct start_req_t *startp;
+    struct val_t *valp;
+    struct data_t data = {};
+    u64 ts;
+    struct request *req = cmd->request;
+
+    //bpf_trace_printk("[Andy]trace_end flag: %x\\n", req->cmd_flags);
```
Comment (netedwardwu) [inline tools/scsisnoop.py:105]: ?

**Review by netedwardwu — COMMENTED** [review]
I have tested on ARM64 kernel 5.10, and these tools are good.
But I didn't check the accuracy of the output data yet.

And leave some suggestions after code review.

File: tools/scsisnoop.py (hunk 8 above)
Comment (weibang6liu) [reply tools/scsisnoop.py:105]: OK, I will check again and remove useless annotate.

**Comment by yonghong-song** [comment]
@weibang6liu Please submit these four tools as four different pull requests, so they can be easily reviewed. Also a dedicated pull request for a single tool helps people go through the commit itself. I suggest to submit scsilatency and scsisnoop first. After these two are settled, you can submit f2fsdist and f2fsslower.

**Comment by weibang6liu** [comment]
> 



> @weibang6liu Please submit these four tools as four different pull requests, so they can be easily reviewed. Also a dedicated pull request for a single tool helps people go through the commit itself. I suggest to submit scsilatency and scsisnoop first. After these two are settled, you can submit f2fsdist and f2fsslower.

OK，that sounds pretty good. I will follow the suggestions.

**Pull request closed without merging**